File: src/main/java/me/kuku/yuq/event/GroupManagerEvent.java
package me.kuku.yuq.event;

import com.IceCreamQAQ.Yu.annotation.Event;
import com.IceCreamQAQ.Yu.annotation.EventListener;
import com.IceCreamQAQ.Yu.cache.EhcacheHelp;
import com.IceCreamQAQ.Yu.job.JobManager;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import com.icecreamqaq.yuq.FunKt;
import com.icecreamqaq.yuq.event.*;
import com.icecreamqaq.yuq.message.*;
import me.kuku.pojo.Result;
import me.kuku.utils.DateTimeFormatterUtils;
import me.kuku.utils.OkHttpUtils;
import me.kuku.yuq.entity.*;
import me.kuku.yuq.logic.BaiduAiLogic;
import me.kuku.yuq.logic.BaiduAiPojo;

import javax.inject.Inject;
import javax.inject.Named;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Random;

@EventListener
public class GroupManagerEvent {
	@Inject
	private GroupService groupService;
	@Inject
	private QqService qqService;
	@Inject
	private MessageService messageService;
	@Inject
	private JobManager jobManager;
	@Inject
	@Named("CommandCountOnTime")
	public EhcacheHelp<Integer> eh;
	@Inject
	private BaiduAiLogic baiduAiLogic;
	@Inject
	private ConfigService configService;

	@Event(weight = Event.Weight.high)
	public void status(com.IceCreamQAQ.Yu.event.events.Event e){
		Long group = null;
		Message message = null;
		if (e instanceof GroupMemberEvent){
			group = ((GroupMemberEvent) e).getGroup().getId();
		}else if (e instanceof GroupMemberRequestEvent){
			group = ((GroupMemberRequestEvent) e).getGroup().getId();
		}else if (e instanceof GroupRecallEvent){
			group = ((GroupRecallEvent) e).getGroup().getId();
		}else if (e instanceof GroupMessageEvent){
			GroupMessageEvent groupMessageEvent = (GroupMessageEvent) e;
			group = groupMessageEvent.getGroup().getId();
			message = groupMessageEvent.getMessage();
		}
		if (group == null) return;
		GroupEntity groupEntity = groupService.findByGroup(group);
		boolean status = true;
		if (message != null) {
			List<String> list = message.toPath();
			if (list.size() == 2) {
				String pa = list.get(1);
				if ("kukubot".equals(list.get(0)) && pa.equals("开") || pa.equals("关")) {
					status = false;
				}
			}
		}
		if (groupEntity != null && Boolean.TRUE.equals(groupEntity.getStatus())){
			status = false;
		}
		if (status){
			e.setCancel(true);
		}
	}

	@Event
	public void qa(GroupMessageEvent e){
		GroupEntity groupEntity = groupService.findByGroup(e.getGroup().getId());
		if (groupEntity == null) return;
		Message message = e.getMessage();
		if (message.toPath().size() == 0) return;
		if ("删问答".equals(message.toPath().get(0))) return;
		String str;
		try {
			str = Message.Companion.firstString(message);
		}catch (IllegalStateException ex){
			return;
		}
		JSONArray qaJsonArray = groupEntity.getQaJson();
		List<String> resultList = new ArrayList<>();
		for (int i = 0; i < qaJsonArray.size(); i++){
			JSONObject jsonObject = qaJsonArray.getJSONObject(i);
			String type = jsonObject.getString("type");
			String q = jsonObject.getString("q");
			boolean status = false;
			if ("ALL".equals(type)){
				if (str.equals(q)) status = true;
			}else if (str.contains(jsonObject.getString("q"))) status = true;
			if (status){
				String a = jsonObject.getString("a");
				resultList.add(a);
			}
		}
		int size = resultList.size();
		if (resultList.size() != 0){
			String a = resultList.get(new Random().nextInt(size));
			e.getGroup().sendMessage(Message.Companion.toMessageByRainCode(a));
		}
	}

	@Event
	public void groupMemberRequest(GroupMemberRequestEvent e){
		GroupEntity groupEntity = groupService.findByGroup(e.getGroup().getId());
		if (groupEntity == null) return;
		if (groupEntity.getAutoReview() != null && groupEntity.getAutoReview()){
			boolean status = true;
			JSONArray blackJsonArray = groupEntity.getBlackJson();
			for (int i = 0; i < blackJsonArray.size(); i++){
				long black = blackJsonArray.getLong(i);
				if (black == e.getQq().getId()){
					status = false;
					break;
				}
			}
			e.setAccept(status);
			e.cancel = true;
		}
	}

	@Event
	public void groupMemberLeave(GroupMemberLeaveEvent e){
		long qq = e.getMember().getId();
		long group = e.getGroup().getId();
		GroupEntity groupEntity = groupService.findByGroup(group);
		QqEntity qqEntity = groupEntity.getQq(qq);
		String msg;
		if (e instanceof GroupMemberLeaveEvent.Leave) {
			String name = e.getMember().getName();
			if (groupEntity.getLeaveGroupBlack() != null && groupEntity.getLeaveGroupBlack()) {
				JSONArray blackJsonArray = groupEntity.getBlackJson();
				blackJsonArray.add(String.valueOf(qq));
				groupEntity.setBlackJson(blackJsonArray);
				groupService.save(groupEntity);
				msg = "刚刚，" + name + "（" + qq + "）退群了，已加入本群黑名单！！";
			} else msg = "刚刚，" + name + "（" + qq + "）离开了我们！！";
			List<MessageEntity> messageList = messageService.findByQqEntityAndGroupEntityOrderByDateDesc(qqEntity, groupEntity);
			Message finallyMessage;
			if (messageList.size() == 0) {
				finallyMessage = Message.Companion.toMessage("他好像还没有说过话！！");
			} else {
				MessageEntity messageEntity = messageList.get(0);
				msg += "\n尽管他就这么的走了，但是我们仍然不要忘记他在[" +
						DateTimeFormatterUtils.format(messageEntity.getDate().getTime(), "yyyy-MM-dd HH:mm:ss") +
						"]说的最后一句话：";
				finallyMessage = Message.Companion.toMessageByRainCode(messageEntity.getContent());
			}
			e.getGroup().sendMessage(Message.Companion.toMessage(msg));
			e.getGroup().sendMessage(finallyMessage);
		}
		qqService.delete(qqEntity);
	}

	@Event
	public void groupMemberJoin(GroupMemberJoinEvent e) {
		long group = e.getGroup().getId();
		long qq = e.getMember().getId();
		GroupEntity groupEntity = groupService.findByGroup(group);
		QqEntity qqEntity = groupEntity.getQq(qq);
		if (Boolean.valueOf(true).equals(groupEntity.getKickWithoutSpeaking())){
			e.getGroup().sendMessage(FunKt.getMif().at(qq).plus("请尽快发言哦，进群5分钟未发言将会被移出本群。"));
			jobManager.registerTimer(() -> {
				List<MessageEntity> list = messageService.findByQqEntityAndGroupEntityOrderByDateDesc(qqEntity, groupEntity);
				if (list.size() == 0){
					e.getMember().kick("进群未发言踢出");
					e.getGroup().sendMessage(Message.Companion.toMessage(qq + "未发送消息，已移出本群"));
				}
			}, 1000 * 60 * 1000);
		}
	}

	@Event
	public void inter(GroupMessageEvent e) {
		GroupEntity groupEntity = groupService.findByGroup(e.getGroup().getId());
		if (groupEntity == null) return;
		Message message = e.getMessage();
		String str;
		try {
			str = Message.Companion.firstString(message);
		}catch (IllegalStateException ex){
			str = null;
		}
		if (str != null){
			JSONArray interceptJsonArray = groupEntity.getInterceptJson();
			for (int i = 0; i < interceptJsonArray.size(); i++){
				String intercept = interceptJsonArray.getString(i);
				if (str.contains(intercept)){
					e.cancel = true;
					break;
				}
			}
		}
		if (!e.getGroup().getBot().isAdmin()) return;
		ConfigEntity configEntity = configService.findByType(ConfigType.BAIDU_AI);
		BaiduAiPojo baiduAiPojo = null;
		if (configEntity != null) baiduAiPojo = configEntity.getConfigParse(BaiduAiPojo.class);
		JSONArray violationJsonArray = groupEntity.getViolationJson();
		int code = 0;
		String vio = null;
		List<Image> images = new ArrayList<>();
		out:for (int i = 0; i < violationJsonArray.size(); i++){
			String violation = violationJsonArray.getString(i);
			String nameCard = e.getSender().getNameCard();
			if (nameCard.contains(violation)) {
				code = 3;
				vio = violation;
				break;
			}
			for (MessageItem item: message.getBody()){
				if (item instanceof Text){
					Text text = (Text) item;
					if (text.getText().contains(violation)) code = 1;
				}else if (item instanceof Image){
					if (baiduAiPojo != null) {
						try {
							Image image = (Image) item;
							images.add(image);
							Result<String> result = baiduAiLogic.ocrGeneralBasic(baiduAiPojo, Base64.getEncoder().encodeToString(OkHttpUtils.getBytes(image.getUrl())));
							if (result.isSuccess()) {
								if (result.getData().contains(violation)) code = 1;
							}
						} catch (Exception ignore) {
						}
					}
				}else if (item instanceof XmlEx){
					XmlEx xmlEx = (XmlEx) item;
					if (xmlEx.getValue().contains(violation)) code = 1;
				}else if (item instanceof JsonEx){
					JsonEx jsonEx = (JsonEx) item;
					if (jsonEx.getValue().contains(violation)) code = 1;
				}
				if (code != 0){
					vio = violation;
					break out;
				}
			}
		}

		if (baiduAiPojo != null && Boolean.TRUE.equals(groupEntity.getPornImage())) {
			for (Image image : images) {
				boolean b = false;
				try {
					b = baiduAiLogic.antiPornImage(baiduAiPojo, Base64.getEncoder().encodeToString(OkHttpUtils.getBytes(image.getUrl())));
					if (b) code = 2;
				} catch (IOException ignore) {
				}
			}
		}
		if (code != 0){
			StringBuilder sb = new StringBuilder();
			if (code == 2) sb.append("检测到色情图片。").append("\n");
			else if (code == 1) sb.append("检测到违规词\"").append(vio).append("\"。").append("\n");
			else sb.append("检测到违规去群名片\"").append(vio).append("\"。").append("\n");
			sb.append("您已被禁言！");
			e.getSender().ban(60 * 30);
			e.getGroup().sendMessage(FunKt.getMif().at(e.getSender().getId()).plus(sb.toString()));
		}
	}

}

File: src/main/kotlin/me/kuku/yuq/event/GroupManagerEvent.kt
package me.kuku.yuq.event

import com.IceCreamQAQ.Yu.annotation.Event
import com.IceCreamQAQ.Yu.annotation.EventListener
import com.icecreamqaq.yuq.event.*
import com.icecreamqaq.yuq.message.Message.Companion.firstString
import com.icecreamqaq.yuq.message.Message.Companion.toCodeString
import com.icecreamqaq.yuq.message.Message.Companion.toMessageByRainCode
import com.icecreamqaq.yuq.message.buildMessage
import com.icecreamqaq.yuq.mif
import me.kuku.utils.DateTimeFormatterUtils
import me.kuku.utils.JobManager
import me.kuku.utils.MyUtils
import me.kuku.yuq.controller.toStatus
import me.kuku.yuq.entity.*
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.CacheManager
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.concurrent.ConcurrentHashMap

@EventListener
@Component
class GroupManagerEvent (
    private val groupService: GroupService,
    private val qqGroupConfigService: QqGroupConfigService,
    private val messageService: MessageService,
    @Suppress("SpringJavaInjectionPointsAutowiringInspection") cacheManager: CacheManager,
    @Value("\${yuq.art.master}") private val master: String
) {

    private val lastMessage = ConcurrentHashMap<Long, String>()
    private val lastQq = ConcurrentHashMap<Long, Long>()
    private val lastRepeatMessage = ConcurrentHashMap<Long, String>()

    private val cache = cacheManager.getCache("CommandLimit")!!

    private val verifyMap = ConcurrentHashMap<String, Boolean>()

    @Event
    @Transactional
    fun inter(e: GroupMessageEvent) {
        val group = e.group
        val groupNum = group.id
        val sender = e.sender
        val qq = sender.id
        val groupEntity = groupService.findByGroup(groupNum) ?: return
        val qqEntity = groupEntity.get(qq) ?: return
        val codeString = e.message.toCodeString()
        val config = groupEntity.config
        val prohibitedWords = config.prohibitedWords
        for (prohibitedWord in prohibitedWords) {
            if (codeString.contains(prohibitedWord)) {
                kotlin.runCatching {
                    sender.ban(60 * 10)
                    val qqGroupEntity = qqGroupConfigService.findByGroupAndQq(groupNum, qq) ?: QqGroupConfigEntity().also {
                        it.groupEntity = groupEntity
                        it.qqEntity = qqEntity
                    }
                    group.sendMessage(mif.at(qq).plus("""
                        您已触发违禁词"$prohibitedWord"，您已被禁言10分钟。
                        您已违规${qqGroupEntity.config.violationsNum}次
                    """.trimIndent()))
                    qqGroupEntity.config.violationsNum += 1
                    qqGroupConfigService.save(qqGroupEntity)
                    e.cancel = true
                    return
                }
            }
        }
    }

    @Event(weight = Event.Weight.lowest)
    fun interceptor(e: GroupMessageEvent) {
        val first = e.message.toPath().getOrNull(0) ?: return
        val groupEntity = groupService.findByGroup(e.group.id) ?: return
        if (groupEntity.config.interceptList.contains(first)) {
            e.cancel = true
        }
    }

    @Event(weight = Event.Weight.lowest)
    fun commandLine(e: GroupMessageEvent) {
        val first = e.message.toPath().getOrNull(0) ?: return
        val groupEntity = groupService.findByGroup(e.group.id) ?: return
        val ss = groupEntity.config.commandLimitList
        for (s in ss) {
            if (s.command == first) {
                val key = if (s.type == CommandLimitType.GROUP) "group${e.group.id}$first"
                else "group${e.group.id}qq${e.sender.id}$first"
                var count = cache.get(key)?.get() as? Int ?: 0
                count += 1
                if (count > s.limit) {
                    e.cancel = true
                    return
                }
                cache.put(key, count)
            }
        }
    }

    @Event(weight = Event.Weight.low)
    fun repeat(e: GroupMessageEvent) {
        val group = e.group
        val groupNum = group.id
        val groupEntity = groupService.findByGroup(groupNum) ?: return
        if (groupEntity.config.repeat == Status.ON) {
            val qq = e.sender.id
            val nowMsg = e.message.toCodeString()
            if (lastMessage.containsKey(groupNum)) {
                val oldMsg = lastMessage[groupNum]
                if (oldMsg == nowMsg && nowMsg != lastRepeatMessage[groupNum] && lastQq[groupNum] != qq) {
                    lastRepeatMessage[groupNum] = nowMsg
                    group.sendMessage(e.message)
                }
            }
            lastMessage[groupNum] = nowMsg
            lastQq[groupNum] = qq
        }
    }

    @Event
    fun groupSwitch(e: GroupMessageEvent) {
        val ss = kotlin.runCatching {
            e.message.firstString()
        }.getOrNull() ?: return
        val group = e.group
        val groupId = group.id
        val groupEntity = groupService.findByGroup(groupId) ?: return
        if (ss in listOf("bot开", "bot关")) {
            val sss = ss.contains("开")
            groupEntity.config.switch = sss.toStatus()
            groupService.save(groupEntity)
            group.sendMessage("bot${if (sss) "开启" else "关闭"}成功")
        } else {
            if (groupEntity.config.switch == Status.OFF) {
                e.cancel = true
            }
        }
    }

    @Event
    fun memberLeave(e: GroupMemberLeaveEvent) {
        val group = e.group
        val groupNum = group.id
        val member = e.member
        val groupEntity = groupService.findByGroup(groupNum) ?: return
        if (groupEntity.config.leaveToBlack == Status.ON) {
            groupEntity.config.blackList.add(member.id)
            groupService.save(groupEntity)
            group.sendMessage("""
                ${member.nameCardOrName()}离开了我们，他（她）已被加入到黑名单
            """.trimIndent())
        } else {
            group.sendMessage("""
                ${member.nameCardOrName()}离开了我们
            """.trimIndent())
        }
        val messageEntityList = messageService.findByGroupAndQqOrderByIdDesc(groupNum, member.id)
        if (messageEntityList.isEmpty()) {
            group.sendMessage("""
                他（她）好像还没有说过话，那没事了
            """.trimIndent())
        } else {
            val messageEntity = messageEntityList[0]
            group.sendMessage("""
                但是我们永远也不要忘记他（她）在群里的${messageEntity.createDate.format(DateTimeFormatterUtils.creat("yyyy-MM-dd HH:mm:ss"))}说的最后一句话
            """.trimIndent())
            group.sendMessage(messageEntity.content.toMessageByRainCode())
        }
    }

    @Event
    fun newRequest(e: NewRequestEvent) {
        val qq: Long
        val groupEntity = when (e) {
            is GroupInviteEvent -> {
                qq = e.qq.id
                val id = e.group.id
                groupService.findByGroup(id)
            }
            is GroupMemberRequestEvent -> {
                qq = e.qq.id
                val id = e.group.id
                groupService.findByGroup(id)
            }
            else -> return
        } ?: return
        val blackList = groupEntity.config.blackList
        if (blackList.contains(qq)) {
            e.accept = false
            e.rejectMessage = "你是黑名单用户，无法加入本群"
            e.cancel = true
        } else {
            e.accept = true
            e.cancel = true
        }
    }

    @Event
    fun friend(e: NewFriendRequestEvent) {
        if (e.message == master) {
            e.accept = true
        } else {
            e.accept = false
            e.rejectMessage = "验证信息不正确"
        }
        e.cancel = true
    }

    @Event
    fun verification(e: GroupMemberJoinEvent) {
        val group = e.group
        val member = e.member
        val qq = e.member.id
        val groupEntity = groupService.findByGroup(group.id) ?: return
        if (groupEntity.config.entryVerification == Status.ON) {
            val captchaCode = MyUtils.randomLetter(4)
            val message = buildMessage {
                at(qq)
                line()
                textLine("欢迎加入本群，请您发送验证码，如若在5分钟内未发送，将会被移除本群")
                text("验证码为$captchaCode")
            }
            group.sendMessage(message)
            val key = group.id.toString() + qq + captchaCode
            verifyMap[key] = false
            JobManager.delay(1000 * 60 * 5) {
                val status = verifyMap[key] ?: return@delay
                verifyMap.remove(key)
                if (!status) {
                    member.kick("超时未验证")
                    group.sendMessage("${qq}超时未验证，已被移出本群")
                }
            }
        }
    }

    @Event
    fun verify(e: GroupMessageEvent) {
        val group = e.group.id
        val qq = e.sender.id
        val captchaCode = kotlin.runCatching {
            e.message.firstString()
        }.getOrNull() ?: return
        val key = group.toString() + qq + captchaCode
        verifyMap.remove(key)
    }
}

